File: core/src/main/java/pcb/uwu/unit/base/AmountOfSubstanceUnit.java
package pcb.uwu.unit.base;

import pcb.uwu.core.BaseUnit;
import pcb.uwu.core.BigDecimalAmount;
import pcb.uwu.core.UnitCounter;

import java.util.function.Function;

public class AmountOfSubstanceUnit extends BaseUnit {
	private final UnitCounter unitCounter;

	public AmountOfSubstanceUnit(String symbol, String singularName, String pluralName,
					  Function<BigDecimalAmount, BigDecimalAmount> translationToCanonical,
					  Function<BigDecimalAmount, BigDecimalAmount> translationFromCanonical) {

		super(1, symbol, singularName, pluralName, translationToCanonical, translationFromCanonical);

		unitCounter = new UnitCounter(this);
	}

	@Override
	public Class<? extends BaseUnit> getBaseUnitType() {
		return AmountOfSubstanceUnit.class;
	}

	@Override
	public UnitCounter getUnitCounter() {
		return unitCounter;
	}
}

File: core/src/main/java/pcb/uwu/unit/base/AmountOfSubstanceUnit.kt
package pcb.uwu.unit.base

import pcb.uwu.core.Amount
import pcb.uwu.core.BaseUnit
import pcb.uwu.core.UnitCounter
import java.util.function.Function

open class AmountOfSubstanceUnit(symbol: String,
                                 singularName: String,
                                 pluralName: String,
                                 translationToCanonical: Function<Amount, Amount>,
                                 translationFromCanonical: Function<Amount, Amount>)
    : BaseUnit(precedence = 1,
               symbol = symbol,
               singularName = singularName,
               pluralName = pluralName,
               translationToCanonical = translationToCanonical,
               translationFromCanonical = translationFromCanonical)
{
    override val unitCounter
        get() = UnitCounter(this)
    override val baseUnitType = AmountOfSubstanceUnit::class.java
}